{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import warnings\n",
    "import seaborn as sns\n",
    "from scipy.stats import gaussian_kde\n",
    "import matplotlib.colors as clr\n",
    "from matplotlib.patches import Polygon\n",
    "import ucell\n",
    "\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "zissou = [\n",
    "    \"#3A9AB2\",\n",
    "    \"#6FB2C1\",\n",
    "    \"#91BAB6\",\n",
    "    \"#A5C2A3\",\n",
    "    \"#BDC881\",\n",
    "    \"#DCCB4E\",\n",
    "    \"#E3B710\",\n",
    "    \"#E79805\",\n",
    "    \"#EC7A05\",\n",
    "    \"#EF5703\",\n",
    "    \"#F11B00\",\n",
    "]\n",
    "\n",
    "colormap = clr.LinearSegmentedColormap.from_list(\"Zissou\", zissou)\n",
    "colormap_r = clr.LinearSegmentedColormap.from_list(\"Zissou\", zissou[::-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(\"../data/adata/timecourse.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "batches = {\n",
    "    # \"day6_SI\": {\"x\": 6200, \"y\": 6200},\n",
    "    # \"day6_SI_r2\": {\"x\": 5800, \"y\": 5500},\n",
    "    # \"day8_SI_Ctrl\": {\"x\": 2400, \"y\": 2400},\n",
    "    # \"day8_SI_r2\": {\"x\": 3200, \"y\": 1500},\n",
    "    # \"day30_SI\": {\"x\": 6400, \"y\": 2400},\n",
    "    # \"day30_SI_r2\": {\"x\": 6200, \"y\": 6200},\n",
    "    \"day90_SI\": {\"x\": 2400, \"y\": 2400},\n",
    "    \"day90_SI_r2\": {\"x\": 1200, \"y\": 6200},\n",
    "}\n",
    "\n",
    "# subset anndata to the day 90 timepoint to make the calculation faster\n",
    "adata = adata[adata.obs.batch.isin(batches.keys())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def import_kurd_et_al_signature(\n",
    "    p_cutoff=0.01,\n",
    "    lfc_cutoff=1,\n",
    "):\n",
    "    \"\"\"\n",
    "    Early precursors and molecular determinants of tissue-resident memory\n",
    "    CD8+ T lymphocytes revealed by single-cell RNA sequencing\n",
    "    -----------------------------------------------------\n",
    "    Table S5 contains DE genes from cluster 3 and cluster 29\n",
    "    - cluster 3 is the more stem like populations\n",
    "    - cluster 29 the more effector like populations\n",
    "    \"\"\"\n",
    "    df = pd.read_excel(\"../data/signatures/kurd.xlsx\", skiprows=1)\n",
    "    df = df[df[\"log10pval\"] < np.log10(p_cutoff)]\n",
    "    up = df[df[\"log2fold\"] > lfc_cutoff][\"Gene\"].to_list()\n",
    "    down = df[df[\"log2fold\"] < lfc_cutoff][\"Gene\"].to_list()\n",
    "    signatures = {\n",
    "        \"cluster_3\": up,\n",
    "        \"cluster_29\": down,\n",
    "    }\n",
    "    return signatures\n",
    "\n",
    "\n",
    "# read signatures\n",
    "def import_milner_et_al_signatures():\n",
    "    \"\"\"\n",
    "    Heterogenous Populations of Tissue-Resident CD8+ T Cells Are\n",
    "    Generated in Response to Infection and Malignancy\n",
    "    --------------------------------------------------\n",
    "    \"\"\"\n",
    "    s = {\n",
    "        \"Blimp1\": pd.read_csv(\"../data/signatures/blimp.txt\", header=None)[0].to_list(),\n",
    "        \"Id3\": pd.read_csv(\"../data/signatures/id3.txt\", header=None)[0].to_list(),\n",
    "    }\n",
    "    return s\n",
    "\n",
    "\n",
    "# Python code to merge dict using update() method\n",
    "def Merge(dict1, dict2):\n",
    "    res = {**dict1, **dict2}\n",
    "    return res\n",
    "\n",
    "\n",
    "signatures = Merge(import_kurd_et_al_signature(), import_milner_et_al_signatures())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def scatter_with_gaussian_kde_weights(ax, x, y, weights, exp, **kwargs):\n",
    "    \"\"\"\n",
    "    Plots a scatter plot colored by gaussian kde estimates.\n",
    "\n",
    "    Parameters:\n",
    "    - ax (matplotlib ax): The ax on which to plot the scatter plot.\n",
    "    - x (np.array): The x values to perform the gaussian kde and scattering on.\n",
    "    - y (np.array): The y values to perform the gaussian kde and scattering on.\n",
    "    - **kwargs: Additional keyword arguments to pass to the scatter function.\n",
    "\n",
    "    Returns:\n",
    "    - None\n",
    "    \"\"\"\n",
    "\n",
    "    xy = np.vstack([x, y])\n",
    "    z = gaussian_kde(xy, weights=weights**exp)(xy)\n",
    "\n",
    "    ax.scatter(x, y, c=z, **kwargs)\n",
    "\n",
    "\n",
    "# Custom biexponential transformation. Returns x coordinates that have been transformed\n",
    "def transformation(x, a=0.1, b=0.1, c=0.5, d=2.5, f=4, w=1):\n",
    "    x = np.array(x)\n",
    "    return a * np.exp(b * ((x - w))) - c * np.exp(-d * (x - w)) + f\n",
    "\n",
    "\n",
    "def draw_gates(ax, gates, transformation, type=\"edge\"):\n",
    "    \"\"\"\n",
    "    Draws IMAP gates on a matplotlib ax.\n",
    "\n",
    "    Parameters:\n",
    "    - ax (matplotlib ax): The ax on which to draw the gates.\n",
    "    - gates (dict): A dictionary containing the gates to draw.\n",
    "    - transformation (function): A function to transform the x values of the gates.\n",
    "    - type (str): The type of gate to draw. Can be \"fill\" or \"edge\".\n",
    "\n",
    "    Returns:\n",
    "    - None\n",
    "    \"\"\"\n",
    "    for gate in gates:\n",
    "        # Apply transformation to x values\n",
    "        points = [\n",
    "            [transformation(element[0])] + element[1:]\n",
    "            for element in gates[gate][\"edges\"]\n",
    "        ]\n",
    "\n",
    "        if type == \"fill\":\n",
    "            p = Polygon(points, facecolor=gates[gate][\"fill\"], edgecolor=\"none\")\n",
    "            ax.add_patch(p)\n",
    "        elif type == \"edge\":\n",
    "            p = Polygon(points, facecolor=\"none\", edgecolor=\"#222222\")\n",
    "            ax.add_patch(p)\n",
    "\n",
    "            ax.text(\n",
    "                transformation(gates[gate][\"label_position\"][\"x\"]),\n",
    "                gates[gate][\"label_position\"][\"y\"],\n",
    "                gate,\n",
    "                fontsize=6,\n",
    "                color=\"#222222\",\n",
    "            )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Coordinates of the gates\n",
    "gates = {\n",
    "    \"Top\": {\n",
    "        \"edges\": [\n",
    "            [0.15, 0.5],\n",
    "            [0.6, 0.7],\n",
    "            [0.8, 0.7],\n",
    "            [0.8, 1.03],\n",
    "            [0.15, 1.03],\n",
    "        ],\n",
    "        \"label_position\": {\"x\": 0.16, \"y\": 0.9},\n",
    "        \"fill\": \"#3A9AB224\",\n",
    "        \"stroke\": \"#3A9AB2\",\n",
    "    },\n",
    "    \"Crypt\": {\n",
    "        \"edges\": [\n",
    "            [0.15, 0.48],\n",
    "            [0.6, 0.68],\n",
    "            [0.8, 0.68],\n",
    "            [0.8, 0.25],\n",
    "            [0.2, 0],\n",
    "            [0.15, 0],\n",
    "        ],\n",
    "        \"label_position\": {\"x\": 0.16, \"y\": 0.05},\n",
    "        \"fill\": \"#F11B0024\",\n",
    "        \"stroke\": \"#F11B00\",\n",
    "    },\n",
    "    \"Muscularis\": {\n",
    "        \"edges\": [[0.22, 0], [0.8, 0.23], [6, 0.23], [6, 0], [0.22, 0]],\n",
    "        \"label_position\": {\"x\": 0.6, \"y\": 0.05},\n",
    "        \"fill\": \"#BDC88135\",\n",
    "        \"stroke\": \"#BDC881\",\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create subplots\n",
    "def plot_imaps_ucell(\n",
    "    adata,\n",
    "    batches,\n",
    "    genes,\n",
    "    ax_ticks=[0.15, 0.3, 0.6, 1, 6],\n",
    "    transformation=transformation,\n",
    "    gates=gates,\n",
    "    dpi=100,\n",
    "    exp=2,\n",
    "):\n",
    "    \"\"\"\n",
    "    Plots IMAPs weighted by a signature defined by \"genes\".\n",
    "\n",
    "    Parameters:\n",
    "    - adata (anndata): The anndata object containing the cells for IMAP plotting.\n",
    "    - batches (list): A list of batches to plot.\n",
    "    - genes (list): The gene signature to weight the IMAP by\n",
    "    - ax_ticks (list): The x-axis ticks to use.\n",
    "    - transformation (function): The transformation to apply to the x-axis.\n",
    "    - gates (dict): The gates to draw on the IMAP.\n",
    "    - dpi (int): The dpi of the plot.\n",
    "    - exp (int): The exponent to use for the gaussian kde.\n",
    "\n",
    "    Returns:\n",
    "    - None\n",
    "    \"\"\"\n",
    "\n",
    "    fig = plt.figure(figsize=(3 * len(genes), 3 * len(batches)), dpi=dpi)\n",
    "\n",
    "    # Apply transformation\n",
    "    adata.obs[\"epithelial_distance_transformed\"] = transformation(\n",
    "        adata.obs[\"epithelial_distance\"]\n",
    "    )\n",
    "\n",
    "    for col, gene in enumerate(genes):\n",
    "        print(\"Plotting value: \" + str(gene))\n",
    "        # Iterate over batches\n",
    "        for i, bt in enumerate(batches):\n",
    "            sub_adata = adata[adata.obs[\"batch\"] == bt]\n",
    "\n",
    "            if gene == \"Distribution\":\n",
    "                gene_expression = np.ones(len(sub_adata))\n",
    "            else:\n",
    "                gene_expression = np.array(sub_adata.obs[gene])\n",
    "\n",
    "            ax = fig.add_subplot(len(batches), len(genes), i * len(genes) + 1 + col)\n",
    "\n",
    "            # Draw gates filled in background\n",
    "            draw_gates(ax, gates=gates, transformation=transformation, type=\"fill\")\n",
    "\n",
    "            # Draw the density lines\n",
    "            sns.kdeplot(\n",
    "                data=sub_adata.obs,\n",
    "                x=\"epithelial_distance_transformed\",\n",
    "                y=\"crypt_villi_axis\",\n",
    "                ax=ax,\n",
    "                weights=gene_expression,\n",
    "                color=\"#444444\",\n",
    "                linewidths=0.5,\n",
    "            )\n",
    "\n",
    "            # Colored scatter plot\n",
    "            scatter_with_gaussian_kde_weights(\n",
    "                ax=ax,\n",
    "                x=sub_adata.obs[\"epithelial_distance_transformed\"],\n",
    "                y=sub_adata.obs[\"crypt_villi_axis\"],\n",
    "                weights=gene_expression,\n",
    "                exp=exp,\n",
    "                s=5,\n",
    "                cmap=\"viridis\" if gene == \"Distribution\" else colormap,\n",
    "            )\n",
    "\n",
    "            # Transform the tick labels and set them\n",
    "            ax.set_xticks(transformation(ax_ticks))\n",
    "            ax.set_xticklabels(ax_ticks)\n",
    "\n",
    "            # Label the axes\n",
    "            ax.set_xlabel(\"Epithelial Axis\")\n",
    "            ax.set_ylabel(f\"{bt}\\nCrypt-Villi Axis\")\n",
    "\n",
    "            ax.set_ylim(-0.02, 1.05)\n",
    "\n",
    "            # Add a title\n",
    "            if i == 0:\n",
    "                ax.set_title(f\"{gene}\")\n",
    "            else:\n",
    "                ax.set_title(f\"\")\n",
    "            draw_gates(ax, gates=gates, transformation=transformation)\n",
    "\n",
    "    fig.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = sns.histplot(np.sum(adata.X > 0, axis=1)[adata.obs[\"Subtype\"] == \"Cd8_T-Cell_P14\"])\n",
    "ax.set_xlabel(\"Number of expressed genes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ucell.add_scores(adata, signatures, maxRank=60, seed=42)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Figure 2j"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_imaps_ucell(\n",
    "    adata[adata.obs[\"Subtype\"] == \"Cd8_T-Cell_P14\"],\n",
    "    [\"day90_SI\"],\n",
    "    genes=[\n",
    "        \"UCell_cluster_3\",\n",
    "        \"UCell_cluster_29\",\n",
    "    ],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extended data figure 2d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_imaps_ucell(\n",
    "    adata[adata.obs[\"Subtype\"] == \"Cd8_T-Cell_P14\"],\n",
    "    [\"day90_SI\"],\n",
    "    genes=[\n",
    "        \"UCell_Id3\",\n",
    "        \"UCell_Blimp1\",\n",
    "    ],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quantification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_p14 = adata[adata.obs[\"Subtype\"] == \"Cd8_T-Cell_P14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def classify_cells(adata, gates, transformation=transformation):\n",
    "    \"\"\"\n",
    "    Classify cells based on the IMAP gates.\n",
    "\n",
    "    Parameters:\n",
    "    - adata (anndata): The anndata object containing the cells to classify.\n",
    "    - gates (dict): A dictionary containing the gates to classify the cells with.\n",
    "    - transformation (function): A function to transform the x values of the gates.\n",
    "\n",
    "    Returns:\n",
    "    - result (geopandas dataframe): A geopandas dataframe containing the classified cells.\n",
    "    \"\"\"\n",
    "    from shapely.geometry import Point\n",
    "    from shapely.geometry.polygon import Polygon\n",
    "    import geopandas as gpd\n",
    "\n",
    "    adata.obs[\"epithelial_distance_transformed\"] = transformation(\n",
    "        adata.obs[\"epithelial_distance\"]\n",
    "    )\n",
    "    adata.obs[\"gate\"] = False\n",
    "\n",
    "    print(\"Creating polygons\")\n",
    "    polygons = {}\n",
    "    for gate in gates:\n",
    "        # Apply transformation to x values\n",
    "        points = [\n",
    "            [transformation(element[0])] + element[1:]\n",
    "            for element in gates[gate][\"edges\"]\n",
    "        ]\n",
    "        polygons[gate] = Polygon(points)\n",
    "    polygons = gpd.GeoSeries(polygons)\n",
    "    gpd_poly = gpd.GeoDataFrame({\"gates\": polygons}, geometry=\"gates\")\n",
    "\n",
    "    print(\"Creating cells\")\n",
    "    cells = gpd.GeoSeries.from_xy(\n",
    "        adata.obs[\"epithelial_distance_transformed\"], adata.obs[\"crypt_villi_axis\"]\n",
    "    )\n",
    "    gpd_cells = gpd.GeoDataFrame({\"cells\": cells}, geometry=\"cells\")\n",
    "\n",
    "    print(\"Joining cells and polygons\")\n",
    "    result = gpd.sjoin(\n",
    "        gpd_cells,\n",
    "        gpd_poly,\n",
    "        how=\"left\",\n",
    "    )\n",
    "    return result\n",
    "\n",
    "\n",
    "def get_mean_expression(adata, genes):\n",
    "    \"\"\"\n",
    "    Calculate the mean expression of specified genes for each batch and gate in the provided AnnData object.\n",
    "\n",
    "    Parameters:\n",
    "    - adata (AnnData): Annotated data matrix with observations (rows) and variables (columns).\n",
    "    - genes (list): A list of gene names for which mean expression is calculated.\n",
    "\n",
    "    Returns:\n",
    "    - pd.DataFrame: A DataFrame containing mean expression values for each gene, batch, and gate.\n",
    "    \"\"\"\n",
    "    import re\n",
    "\n",
    "    keys = genes + [\"batch\", \"gate\"]\n",
    "    df = sc.get.obs_df(adata, keys=keys)\n",
    "    grouped = (\n",
    "        df.groupby([\"batch\", \"gate\"])\n",
    "        .mean()\n",
    "        .join(df.groupby([\"batch\", \"gate\"]).size().rename(\"group_size\"))\n",
    "        .reset_index()\n",
    "    )\n",
    "    grouped[\"Day\"] = [re.findall(r\"\\d+\", b)[0] for b in grouped[\"batch\"]]\n",
    "\n",
    "    # Melt into a longer form\n",
    "    grouped = pd.melt(\n",
    "        grouped,\n",
    "        id_vars=[\"batch\", \"Day\", \"gate\", \"group_size\"],\n",
    "        var_name=\"gene\",\n",
    "        value_name=\"expression\",\n",
    "    )\n",
    "    return grouped\n",
    "\n",
    "\n",
    "def get_scaled_mean_expression(adata, genes):\n",
    "    \"\"\"\n",
    "    Calculate the scaled mean expression of specified genes for each batch and gate in the provided AnnData object.\n",
    "\n",
    "    Parameters:\n",
    "    - adata (AnnData): Annotated data matrix with observations (rows) and variables (columns).\n",
    "    - genes (list): A list of gene names for which scaled mean expression is calculated.\n",
    "\n",
    "    Returns:\n",
    "    - pd.DataFrame: A DataFrame containing scaled mean expression values for each gene, batch, and gate.\n",
    "    \"\"\"\n",
    "    grouped = get_mean_expression(adata=adata, genes=genes)\n",
    "    # Group the DataFrame by gene\n",
    "    gene_groups = grouped.groupby(\"gene\")\n",
    "\n",
    "    # Apply the scaling function to each group\n",
    "    def scale_group(group):\n",
    "        gene_min = group[\"expression\"].min()\n",
    "        gene_max = group[\"expression\"].max()\n",
    "        group[\"expression\"] = (group[\"expression\"] - gene_min) / (gene_max - gene_min)\n",
    "        return group\n",
    "\n",
    "    df_scaled = gene_groups.apply(scale_group)\n",
    "\n",
    "    return df_scaled"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "classification = classify_cells(adata_p14, gates)\n",
    "adata_p14.obs[\"gate\"] = classification[\"index_right\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = get_scaled_mean_expression(adata_p14, [\"UCell_cluster_3\", \"UCell_cluster_29\"])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.pivot(columns=\"gate\", values=\"expression\", index=[\"gene\", \"batch\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sns.FacetGrid(df[df[\"gate\"] != \"Muscularis\"], col=\"gene\", col_wrap=4, sharey=False)\n",
    "g.map_dataframe(\n",
    "    sns.barplot,\n",
    "    x=\"gate\",\n",
    "    y=\"expression\",\n",
    "    capsize=0.2,\n",
    "    edgecolor=\"black\",\n",
    "    linewidth=0.5,\n",
    "    errwidth=1,\n",
    "    color=\"#3A9AB2\",\n",
    ")\n",
    "g.map_dataframe(sns.swarmplot, x=\"gate\", y=\"expression\", hue=\"batch\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
